src/Microsoft.ML.FastTree/GamModelParameters.cs

```diff
@@ -0,0 +1,911 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.ML;
+using Microsoft.ML.Calibrator;
+using Microsoft.ML.Command;
+using Microsoft.ML.CommandLine;
+using Microsoft.ML.Data;
+using Microsoft.ML.Internal.Calibration;
+using Microsoft.ML.Internal.Internallearn;
+using Microsoft.ML.Internal.Utilities;
+using Microsoft.ML.Model;
+using Microsoft.ML.Trainers.FastTree;
+using Microsoft.ML.Trainers.FastTree.Internal;
+using Microsoft.ML.Training;
+
+[assembly: LoadableClass(typeof(GamModelParametersBase.VisualizationCommand), typeof(GamModelParametersBase.VisualizationCommand.Arguments), typeof(SignatureCommand),
+    "GAM Vizualization Command", GamModelParametersBase.VisualizationCommand.LoadName, "gamviz", DocName = "command/GamViz.md")]
+
+namespace Microsoft.ML.Trainers.FastTree
+{
+    /// <summary>
+    /// The base class for GAM Model Parameters.
+    /// </summary>
+    public abstract class GamModelParametersBase : ModelParametersBase<float>, IValueMapper, ICalculateFeatureContribution,
+        IFeatureContributionMapper, ICanSaveInTextFormat, ICanSaveSummary, ICanSaveInIniFormat
+    {
+        private readonly double[][] _binUpperBounds;
+        private readonly double[][] _binEffects;
+        public readonly double Intercept;
+        private readonly int _numFeatures;
+        private readonly ColumnType _inputType;
+        private readonly ColumnType _outputType;
+        // These would be the bins for a totally sparse input.
+        private readonly int[] _binsAtAllZero;
+        // The output value for all zeros
+        private readonly double _valueAtAllZero;
+
+        private readonly int[] _featureMap;
+        private readonly int _inputLength;
+        private readonly Dictionary<int, int> _inputFeatureToDatasetFeatureMap;
+
+        ColumnType IValueMapper.InputType => _inputType;
+        ColumnType IValueMapper.OutputType => _outputType;
+
+        /// <summary>
+        /// Used to determine the contribution of each feature to the score of an example by <see cref="FeatureContributionCalculatingTransformer"/>.
+        /// For Generalized Additive Models (GAM), the contribution of a feature is equal to the shape function for the given feature evaluated at
+        /// the feature value.
+        /// </summary>
+        public FeatureContributionCalculator FeatureContributionClaculator => new FeatureContributionCalculator(this);
```

Comment:
>FeatureContributionCalculator [](start = 15, length = 29)

this class has one internal property and one internal method.
I advice to make it internal instead of public (as class in general and as property of Gam predictor)

Reply:
Can you clarify what you mean here?

Reply:
This thing gonna be visible to any user. In same time if you look on content of this class, it has only internal objects.
Basically during debugging user will see FeatureContributionCalculator, but it will have zero public methods and properties.
We have that thing for a reason, but I doubt it should be public.

Reply:
@artidoro Does this method need to be `public` to work?